### ISSUE TITLE
Clarification Regarding Bounded Wildcards for Generics in map Method for ImList

### ISSUE THREAD

**Issue by Geraldworks** [nus-cs2030/2122-s2#89]
# Description

Hi guys I would like to ask what the difference is between using map1 and map2 class methods. Prof's Lecture example used `ImList.<Number>of(List.<Integer>of(1,2,3)).map(x -> x.hashCode())`

From my understanding of anonymous functions, `x -> x.hashCode()` is saying that we will take every element x in the ImList<Number> and map it to an Integer. Both map1 and map2 class methods in the screenshot below achieve the same purpose. So wouldn't these bounded wildcards seem redundant?

`? super T` would imply that T has everything from its parents because of inheritance, so why would you purposely want something that is a superclass of T, when T itself already does everything. Also, with respect to referencing from the ImList<Number> itself, wouldn't the x in `x -> x.hashCode()` always be referencing T type (for this case Number)? 

As for `? extends R`, when you create an anonymous function, aren't you already deciding on the return type of the mapper function? For instance, `x -> x.hashCode()` is explicitly saying you would want an Integer. So once again, adding the bounded wildcards seem redundant unless you could have different return types.

I think I'm misunderstanding something, but jshell seems to be telling me that either implementations (map1, map2) work fine.. :c

## Topic:

Generics
Bounded Wildcards

## Screenshots

Screenshots (if any)
<img width="428" alt="Screenshot 2022-03-13 at 6 12 16 PM" src="https://user-images.githubusercontent.com/97504394/158054769-b0fdf553-60c2-4cb9-ac37-e4bd5175778e.png">

**Comment by szeying02** [comment]
I think in your example, the wildcard are redundant. However, if you want to equate the list of integers to a number, 
`ImList.<Number> nums = ImList.<Number>of(List.<Integer>of(1,2,3)).map(x -> x.hashCode())`, it wouldn't work in the case of map 1, while it work word in the case of map 2, as a result of the wildcard. Prof talks about this wildcard part in the lecture @ 1:23:30, and the specific `nums` part @ 1:30:51. Hope this helps!

**Comment by Geraldworks** [comment]
<img width="447" alt="Screenshot 2022-03-13 at 7 02 54 PM" src="https://user-images.githubusercontent.com/97504394/158056493-8ed9a6b7-1708-473c-9bfb-69b3202ecb2f.png">
both works though

**Comment by szeying02** [comment]
Oh wait, what about an example like this:

Let T is a subclass of S, and T has this function (say Q q()) which S doesn't have. 
Let Q be a sub class of R. 

`ImList<R> lst = ImList.<R>of(List.<Q>.of(...))`

So in the anonymous function, you input an object of class S, instantiated as x.
`lst.map1(x -> x.q())` --> would give an error I believe, because the compile-time type is S, and thus cannot access the methods of T. 
Whereas, 
`lst.map2(x -> x.q())` --> would work I believe, because the compile-time type is `? super S`, which can call on method `q.()`

I'm not too sure why the return type in pic above works though :( sorry.

**Comment by Geraldworks** [comment]
Yeah its pretty strange isn't it
Quite fascinating HAHA

**Comment by szeying02** [comment]
Yeah, it doesn't make sense to me haha 
at least we made sense of the half of it! :)

**Comment by Geraldworks** [comment]
I was thinking java probably does its own internal typecasting.

Let's say I don't assign `ImList<Number> var` to the output of the map function. It would generate an `ImList<Integer>`, maybe internally it is assigned as $50. when I try to assign this to a new variable `ImList<Number> var = $50`, clearly it wouldn't work. Strangely doing it all in one line would... 

I also used a more general ImList type `ImList<Object>` and I got the same results.

**Comment by szeying02** [comment]
Yeah I was thinking of something along those lines. I tried the same thing using `x -> x.toString()` and typecasting to `ImList<Object>`. It yielded the same results :(

**Comment by hwkchia** [comment]
Sorry, I'm kind of lost in your discussion.  What are you all trying to do again?  Go slowly please?  I'm getting old...

**Comment by Geraldworks** [comment]
Hi Prof!
I wanted to clarify why the map function used in lecture 8 requires the `? extends` or `? super` keywords.
<img width="428" alt="img1" src="https://user-images.githubusercontent.com/97504394/158761063-f9ac2c14-fc1f-439c-adbe-28720ec47483.png">
The image above has "map1" and "map2" methods, with map2 being the more general version of the map function. 

The lecture example used
`ImList<Number> numList = ImList.<Number>of(List.<Integer>of(1,2,3)).map(x -> x.hashCode())`
To show that the generic form can be assigned to `ImList<Number>` type.

However, when I tried this exact same line of code using map1 (the non-generic version), I still am able to assign it to an `ImList<Number>` type.

I tested both "map1" and "map2" in jshell as shown below and did not get any compilation error
<img width="447" alt="img2" src="https://user-images.githubusercontent.com/97504394/158761880-4783f71d-a363-4715-89f9-6f246542fcbb.png">

So it seems strange that an `ImList<Integer>` can be assigned to an `ImList<Number>` for both implementations of the map function and we're not exactly sure why this is the case.

**Comment by hwkchia** [comment]
The confusion lies with type inference.  Let me break it down for you.

First of all
```
ImList<Number> numList = ImList.<Number>of(List.<Integer>of(1,2,3))
```
means that `numList` is a `ImList<Number>`

Let's define a function as
```
Function<Object,Integer> f = x -> x.hashCode()
```
In this case, `numList.map1(f)` will fail, while calling `map2` works due to the generality.

You see that `ImList<Number> thing = numList.map1(x -> x.hashCode())` works because the function has been inferred to `Function<Number,Number>`.  Indeed,
```
Function<Number,Number> f = x -> x.hashCode()
```
is valid.  We are taking a `Number` and calling it's `hashCode` (valid since `Number` is an `Object`).  It will produce an `Integer` but we only need the `Number` part (valid since `Integer` is a `Number`).
